Enable ClearLinux as a supported distro

ganeshmaharaj opened the pull request:
This PR enables ClearLinux as a supported distribution with ceph-deploy. ClearLinux primarily only contains python3 which prompted another commit to add python3 support.

ceph-jenkins commented:
Can one of the admins verify this patch?

HacKanCuBa commented on ceph_deploy/hosts/clear/__init__.py:
```diff
@@ -0,0 +1,23 @@
+    """
+    Select a init system
+
+    Retunrss the name of a init system (upstart, sysvinit ...).
```
Typo in `Retunrss`

HacKanCuBa commented on ceph_deploy/hosts/clear/install.py:
```diff
@@ -0,0 +1,14 @@
+
+    logger.info("Installing storage-cluster bundle")
+    distro.packager.install(
+            packages
```
Careful w/ tabs

ganeshmaharaj replied:
Thought this was all spaces. Reduced it to 4 to match other files.

HacKanCuBa commented on ceph_deploy/hosts/clear/uninstall.py:
```diff
@@ -0,0 +1,41 @@
+    packages = [
+        'storage-cluster',
+        'ceph'
+        ]
```
Extra tab

ganeshmaharaj replied:
i am not able to spot the tab here and flake8 isn't complaining either. Could you explain a bit more on where the extra tab is?

HacKanCuBa replied:
at line 39. it's technically OK though, so no problem.

HacKanCuBa commented on ceph_deploy/hosts/clear/install.py:
```diff
@@ -0,0 +1,14 @@
+    Since clear does not have alternate channels, we will just run the command
+    """
+    logger = distro.conn.logger
+    packages = ['storage-cluster']
```
Differs from packages in uninstall.py, is this correct?

ganeshmaharaj replied:
Yes, storage-cluster is a super-set that contains ceph and other bits. the uninstall list explicitly calls ceph just to be sure that we get them all removed.

HacKanCuBa commented on ceph_deploy/hosts/remotes.py:
```diff
@@ -153,6 +153,10 @@ def remove_whitespace_from_assignments():
+    # Stateless OS like ClearLinux does not have anything populated in /etc.
+    # Create the folder before we write the config file.
+    if not os.path.isdir("/etc/ceph"):
+        os.makedirs("/etc/ceph")
```
DANGEROUS! `os.makedirs` defaults to mode 777, change line to `os.makedirs('/etc/ceph', mode=0o700)`

On another hand, try to keep the quotation marks the files is using (single in this case). Would be nice to change both lines.

HacKanCuBa commented on ceph_deploy/util/pkg_managers.py:
```diff
@@ -427,3 +427,48 @@ def remove(self, packages, **kw):
+    """
+
+    executable = [
+            'swupd',
```
Extra tab

HacKanCuBa commented:
Hi @ganeshmaharaj ! Thanks for your commit!
I'm not an admin but just a contributor, so don't take my word like any more than just a suggestion.

I reviewed your code, yet I haven't tested it, and seems good. Just one detail I marked as dangerous that really needs amending, the rest is mere style. I would recommend running flake8 over your new code.

Cheers!

ganeshmaharaj commented:
@HacKanCuBa no worries. All reviews are welcome. I have cleaned up the code as best as i can addressing your comments. Let me know how they look this time.

HacKanCuBa commented:
> @HacKanCuBa no worries. All reviews are welcome. I have cleaned up the code as best as i can addressing your comments. Let me know how they look this time.

Awesome! I'll be reviewing ASAP :)

HacKanCuBa commented on ceph_deploy/hosts/remotes.py:
```diff
@@ -153,6 +153,10 @@ def remove_whitespace_from_assignments():
+    # Stateless OS like ClearLinux does not have anything populated in /etc.
+    # Create the folder before we write the config file.
+    if not os.path.isdir('/etc/ceph'):
+        os.makedirs('/etc/ceph', mode=0o700)
```
I think we should rewrite this function, because at the end of the function there's a check and an exception if /etc/ceph doesn't exists, which is nonsense now.
I'm not sure I can commit to this branch so here goes mi proposal:

```python
def write_conf(cluster, conf, overwrite):
    """ write cluster configuration to /etc/ceph/{cluster}.conf """
    path = '/etc/ceph/{cluster}.conf'.format(cluster=cluster)

    # Stateless OS like ClearLinux does not have anything populated in /etc.
    # Create the folder before we write the config file.
    if not os.path.isdir('/etc/ceph'):
        os.makedirs('/etc/ceph', mode=0o700)

    if os.path.exists(path):
        tmp_file = tempfile.NamedTemporaryFile('w', dir='/etc/ceph', delete=False)
        err_msg = 'config file %s exists with different content; use ' \
                  '--overwrite-conf to overwrite' % path
        with open(path, 'r') as f:
            old = f.read()
            if old != conf and not overwrite:
                raise RuntimeError(err_msg)
        tmp_file.write(conf)
        tmp_file.close()
        shutil.move(tmp_file.name, path)
    else:
        with open(path, 'w') as f:
            f.write(conf)
    os.chmod(path, 0o644)
```

ganeshmaharaj replied:
@HacKanCuBa  thnks for the tip. Change integrated.

ganeshmaharaj replied:
@HacKanCuBa  thanks. done.

YangLiang3 commented:
@ganeshmaharaj  Could you please solve this confict issues and move it forward?

tchaikov commented:
@ganeshmaharaj the python3 bits in this changeset do not apply anymore. could you let me know if you want update this PR and move it forwards, i'd like to cut a new release in couple weeks. otherwise i will close this PR.

ganeshmaharaj commented:
@tchaikov will work on fixing up the patch this week. Apologies for any delay on my end. Had been on a hiatus from github for a while.

tchaikov commented:
@ganeshmaharaj no worries. sorry for the short notice. the previous maintainer of this project does not work on it anymore.  i feel it would help the community if we can have another release of ceph-deploy in a predictable timeline mainly to address python3 support and the pending PRs.

tchaikov commented on ceph_deploy/hosts/remotes.py:
```diff
@@ -158,26 +158,25 @@ def remove_whitespace_from_assignments():
-    else:
-        err_msg = '/etc/ceph/ does not exist - could not write config'
-        raise RuntimeError(err_msg)
+    os.chmod(path, 0o644)
```
nice cleanup.

tchaikov commented on ceph_deploy/hosts/remotes.py:
```diff
@@ -158,26 +158,25 @@ def remove_whitespace_from_assignments():
-    err_msg = 'config file %s exists with different content; use --overwrite-conf to overwrite' % path
+    # Stateless OS like ClearLinux does not have anything populated in /etc.
+    # Create the folder before we write the config file.
+    if not os.path.isdir('/etc/ceph'):
```
@ganeshmaharaj this changes the existing behavior even on non-ClearLinux distros.

before your change, we bail out if `/etc/ceph` does not exist with the error message of 

> '/etc/ceph/ does not exist - could not write config'

but after your change this directory is always created as long as it does not exist beforehand. i understand Clear Linux tries to avoid placing anything in `/etc`, see https://clearlinux.org/news-blogs/where-etcfstab-clear-linux . does this imply that `swupd` does not create `/etc/ceph` for us when/after installing ceph packages?

if that's the intended behavior of `swupd` and the ceph packages on ClearLinux, how about creating `/etc/ceph` in `Swupd.install()` ? so we don't change the existing behavior on other distros?

ganeshmaharaj replied:
@tchaikov yes, clearlinux does not create `/etc/ceph` when we install packages by default. I will move the folder creation to Swupd.install() and get the changes here back to where they were.

tchaikov commented:
jenkins test this please.
